File: backend/models/input.py
from pydantic import BaseModel
from pydantic.functional_validators import AfterValidator

from typing_extensions import Annotated

def check_location_validity(s: str) -> str:
    assert len(s) != 0
    return s.lower()

def check_property_validaty(v: float) -> float:
    assert v > 0
    return v

Location = Annotated[str, AfterValidator(check_location_validity)]

Property = Annotated[float, AfterValidator(check_property_validaty)]

class BaseInput(BaseModel):
    location_a: Location
    location_b: Location
    servis_type: str = "default"
    extra_services: list[str] = []

class LetterInput(BaseInput):
    ...
    

class BoxProperties(BaseModel):
    width: Property = 1
    height: Property = 1
    length: Property = 1
    weight: Property = 1
    

class BoxInput(BaseInput):
    properties: BoxProperties

File: frontend/yolla/src/app/types/input.ts
// BaseInput type
export type BaseInput = {
  location_a: string;
  location_b: string;
  servis_type?: string; // Default value handled in backend
  extra_services?: string[]; // Default is an empty array
};

// LetterInput type (inherits from BaseInput)
export type LetterInput = BaseInput;

// BoxProperties type
export type BoxProperties = {
  width?: number; // Default value handled in backend
  height?: number; // Default value handled in backend
  length?: number; // Default value handled in backend
  weight?: number; // Default value handled in backend
};

// BoxInput type (inherits from BaseInput and adds properties)
export type BoxInput = BaseInput & {
  properties: BoxProperties;
};
